### ISSUE TITLE
Improve dependency injection

### ISSUE THREAD

**Issue by ycanardeau** [syuilo/Misskey#9881]
## Summary

TL;DR: Please feel free to close this issue if you prefer battle-tested NestJS DI container.

There is a working branch [here](https://github.com/ycanardeau/misskey/tree/improve-dependency-injection).

This branch replaces NestJS DI container with [Yohira](https://github.com/ycanardeau/yohira) DI container. I also changed code and file structures a bit so that they are more readable and understandable. I'm not sure if changing structures is a good idea though considering that https://misskey.io/notes/9b1z035voq will be published on Monday.

ASP.NET Core has a built-int validation feature, which is disabled in production by default, to detect missing and/or circular dependencies in DI at runtime. I [ported Microsoft dependency injection container to TypeScript](https://github.com/ycanardeau/yohira/tree/a66c31fff59b4b75e298082ef447639b9091fcc7/packages/extensions.dependency-injection). You need to *temporarily* set `emitDecoratorMetadata` to `false` to use this feature, otherwise JavaScript throws a cryptic circular dependency error instead. For example, you can detect circular dependencies as follows:
```
Some services are not able to be constructed
Error while validating the service descriptor 'serviceType: AccountUpdateService lifetime: Singleton implCtor: AccountUpdateService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
AccountUpdateService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: AntennaService lifetime: Singleton implCtor: AntennaService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
AntennaService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: AchievementService lifetime: Singleton implCtor: AchievementService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
AchievementService -> CreateNotificationService -> NotificationEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: CreateNotificationService lifetime: Singleton implCtor: CreateNotificationService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
CreateNotificationService -> NotificationEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: CustomEmojiService lifetime: Singleton implCtor: CustomEmojiService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
CustomEmojiService -> EmojiEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DeleteAccountService lifetime: Singleton implCtor: DeleteAccountService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DeleteAccountService -> UserSuspendService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DriveService lifetime: Singleton implCtor: DriveService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: HashtagService lifetime: Singleton implCtor: HashtagService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
HashtagService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: MessagingService lifetime: Singleton implCtor: MessagingService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
MessagingService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NoteCreateService lifetime: Singleton implCtor: NoteCreateService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NoteCreateService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NoteDeleteService lifetime: Singleton implCtor: NoteDeleteService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NoteDeleteService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NotePiningService lifetime: Singleton implCtor: NotePiningService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NotePiningService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NoteReadService lifetime: Singleton implCtor: NoteReadService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NoteReadService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NotificationService lifetime: Singleton implCtor: NotificationService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NotificationService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PollService lifetime: Singleton implCtor: PollService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PollService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ReactionService lifetime: Singleton implCtor: ReactionService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ReactionService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: RelayService lifetime: Singleton implCtor: RelayService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
RelayService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: RoleService lifetime: Singleton implCtor: RoleService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
RoleService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SignupService lifetime: Singleton implCtor: SignupService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SignupService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserBlockingService lifetime: Singleton implCtor: UserBlockingService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserBlockingService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserCacheService lifetime: Singleton implCtor: UserCacheService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserFollowingService lifetime: Singleton implCtor: UserFollowingService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserFollowingService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserListService lifetime: Singleton implCtor: UserListService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserListService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserSuspendService lifetime: Singleton implCtor: UserSuspendService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserSuspendService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UsersChart lifetime: Singleton implCtor: UsersChart': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PerUserReactionsChart lifetime: Singleton implCtor: PerUserReactionsChart': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PerUserReactionsChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PerUserFollowingChart lifetime: Singleton implCtor: PerUserFollowingChart': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PerUserFollowingChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PerUserDriveChart lifetime: Singleton implCtor: PerUserDriveChart': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PerUserDriveChart -> DriveFileEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ChartManagementService lifetime: Singleton implCtor: ChartManagementService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ChartManagementService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: AbuseUserReportEntityService lifetime: Singleton implCtor: AbuseUserReportEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
AbuseUserReportEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: BlockingEntityService lifetime: Singleton implCtor: BlockingEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
BlockingEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ChannelEntityService lifetime: Singleton implCtor: ChannelEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ChannelEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ClipEntityService lifetime: Singleton implCtor: ClipEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ClipEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DriveFileEntityService lifetime: Singleton implCtor: DriveFileEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DriveFileEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: EmojiEntityService lifetime: Singleton implCtor: EmojiEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
EmojiEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: FollowingEntityService lifetime: Singleton implCtor: FollowingEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
FollowingEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: FollowRequestEntityService lifetime: Singleton implCtor: FollowRequestEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
FollowRequestEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: GalleryLikeEntityService lifetime: Singleton implCtor: GalleryLikeEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
GalleryLikeEntityService -> GalleryPostEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: GalleryPostEntityService lifetime: Singleton implCtor: GalleryPostEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
GalleryPostEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: HashtagEntityService lifetime: Singleton implCtor: HashtagEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
HashtagEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: MessagingMessageEntityService lifetime: Singleton implCtor: MessagingMessageEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
MessagingMessageEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ModerationLogEntityService lifetime: Singleton implCtor: ModerationLogEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ModerationLogEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: MutingEntityService lifetime: Singleton implCtor: MutingEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
MutingEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NoteEntityService lifetime: Singleton implCtor: NoteEntityService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: NoteReactionEntityService lifetime: Singleton implCtor: NoteReactionEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NoteReactionEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NotificationEntityService lifetime: Singleton implCtor: NotificationEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NotificationEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PageEntityService lifetime: Singleton implCtor: PageEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PageEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: PageLikeEntityService lifetime: Singleton implCtor: PageLikeEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
PageLikeEntityService -> PageEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SigninEntityService lifetime: Singleton implCtor: SigninEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SigninEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserEntityService lifetime: Singleton implCtor: UserEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserGroupEntityService lifetime: Singleton implCtor: UserGroupEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserGroupEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserGroupInvitationEntityService lifetime: Singleton implCtor: UserGroupInvitationEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserGroupInvitationEntityService -> UserGroupEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserListEntityService lifetime: Singleton implCtor: UserListEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
UserListEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: FlashEntityService lifetime: Singleton implCtor: FlashEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
FlashEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: FlashLikeEntityService lifetime: Singleton implCtor: FlashLikeEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
FlashLikeEntityService -> FlashEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: RoleEntityService lifetime: Singleton implCtor: RoleEntityService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
RoleEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApAudienceService lifetime: Singleton implCtor: ApAudienceService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApAudienceService -> ApPersonService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApDbResolverService lifetime: Singleton implCtor: ApDbResolverService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApDbResolverService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApDeliverManagerService lifetime: Singleton implCtor: ApDeliverManagerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApDeliverManagerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApInboxService lifetime: Singleton implCtor: ApInboxService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApInboxService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApRendererService lifetime: Singleton implCtor: ApRendererService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApResolverService lifetime: Singleton implCtor: ApResolverService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApResolverService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: RemoteUserResolveService lifetime: Singleton implCtor: RemoteUserResolveService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
RemoteUserResolveService -> ApPersonService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApImageService lifetime: Singleton implCtor: ApImageService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApImageService -> ApResolverService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApMentionService lifetime: Singleton implCtor: ApMentionService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApMentionService -> ApResolverService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApNoteService lifetime: Singleton implCtor: ApNoteService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApNoteService -> ApResolverService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApPersonService lifetime: Singleton implCtor: ApPersonService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApPersonService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApQuestionService lifetime: Singleton implCtor: ApQuestionService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApQuestionService -> ApResolverService -> ApRendererService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ClientServerService lifetime: Singleton implCtor: ClientServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ClientServerService -> FlashEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: FeedService lifetime: Singleton implCtor: FeedService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
FeedService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ActivityPubServerService lifetime: Singleton implCtor: ActivityPubServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ActivityPubServerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: NodeinfoServerService lifetime: Singleton implCtor: NodeinfoServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
NodeinfoServerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ServerService lifetime: Singleton implCtor: ServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ServerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: WellKnownServerService lifetime: Singleton implCtor: WellKnownServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
WellKnownServerService -> NodeinfoServerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: GetterService lifetime: Singleton implCtor: GetterService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
GetterService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ChannelsService lifetime: Singleton implCtor: ChannelsService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
ChannelsService -> MainChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: ApiCallService lifetime: Singleton implCtor: ApiCallService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApiCallService -> AuthenticateService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ApiServerService lifetime: Singleton implCtor: ApiServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ApiServerService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: AuthenticateService lifetime: Singleton implCtor: AuthenticateService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
AuthenticateService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SigninApiService lifetime: Singleton implCtor: SigninApiService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SigninApiService -> SigninService -> SigninEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SigninService lifetime: Singleton implCtor: SigninService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SigninService -> SigninEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SignupApiService lifetime: Singleton implCtor: SignupApiService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SignupApiService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: StreamingApiServerService lifetime: Singleton implCtor: StreamingApiServerService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
StreamingApiServerService -> NoteReadService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: MainChannelService lifetime: Singleton implCtor: MainChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
MainChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: AntennaChannelService lifetime: Singleton implCtor: AntennaChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
AntennaChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: ChannelChannelService lifetime: Singleton implCtor: ChannelChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
ChannelChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: GlobalTimelineChannelService lifetime: Singleton implCtor: GlobalTimelineChannelService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
GlobalTimelineChannelService -> RoleService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: HashtagChannelService lifetime: Singleton implCtor: HashtagChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
HashtagChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: HomeTimelineChannelService lifetime: Singleton implCtor: HomeTimelineChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
HomeTimelineChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: HybridTimelineChannelService lifetime: Singleton implCtor: HybridTimelineChannelService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
HybridTimelineChannelService -> RoleService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: LocalTimelineChannelService lifetime: Singleton implCtor: LocalTimelineChannelService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
LocalTimelineChannelService -> RoleService -> UserCacheService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: MessagingChannelService lifetime: Singleton implCtor: MessagingChannelService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
MessagingChannelService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: UserListChannelService lifetime: Singleton implCtor: UserListChannelService': Error: A circular dependency was detected for the service of type 'NoteEntityService'.
UserListChannelService -> NoteEntityService -> UserEntityService -> NoteEntityService
Error while validating the service descriptor 'serviceType: TickChartsProcessorService lifetime: Singleton implCtor: TickChartsProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
TickChartsProcessorService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ResyncChartsProcessorService lifetime: Singleton implCtor: ResyncChartsProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ResyncChartsProcessorService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: CleanChartsProcessorService lifetime: Singleton implCtor: CleanChartsProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
CleanChartsProcessorService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DeleteDriveFilesProcessorService lifetime: Singleton implCtor: DeleteDriveFilesProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DeleteDriveFilesProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportCustomEmojisProcessorService lifetime: Singleton implCtor: ExportCustomEmojisProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportCustomEmojisProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportNotesProcessorService lifetime: Singleton implCtor: ExportNotesProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportNotesProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportFavoritesProcessorService lifetime: Singleton implCtor: ExportFavoritesProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportFavoritesProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportFollowingProcessorService lifetime: Singleton implCtor: ExportFollowingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportFollowingProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportMutingProcessorService lifetime: Singleton implCtor: ExportMutingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportMutingProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportBlockingProcessorService lifetime: Singleton implCtor: ExportBlockingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportBlockingProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ExportUserListsProcessorService lifetime: Singleton implCtor: ExportUserListsProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ExportUserListsProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ImportFollowingProcessorService lifetime: Singleton implCtor: ImportFollowingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ImportFollowingProcessorService -> UserFollowingService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ImportMutingProcessorService lifetime: Singleton implCtor: ImportMutingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ImportMutingProcessorService -> RemoteUserResolveService -> ApPersonService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ImportBlockingProcessorService lifetime: Singleton implCtor: ImportBlockingProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ImportBlockingProcessorService -> UserBlockingService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ImportUserListsProcessorService lifetime: Singleton implCtor: ImportUserListsProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ImportUserListsProcessorService -> UserListService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ImportCustomEmojisProcessorService lifetime: Singleton implCtor: ImportCustomEmojisProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ImportCustomEmojisProcessorService -> CustomEmojiService -> EmojiEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DeleteAccountProcessorService lifetime: Singleton implCtor: DeleteAccountProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DeleteAccountProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DeleteFileProcessorService lifetime: Singleton implCtor: DeleteFileProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DeleteFileProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: CleanRemoteFilesProcessorService lifetime: Singleton implCtor: CleanRemoteFilesProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
CleanRemoteFilesProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: SystemQueueProcessorsService lifetime: Singleton implCtor: SystemQueueProcessorsService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
SystemQueueProcessorsService -> TickChartsProcessorService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: ObjectStorageQueueProcessorsService lifetime: Singleton implCtor: ObjectStorageQueueProcessorsService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
ObjectStorageQueueProcessorsService -> DeleteFileProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: DbQueueProcessorsService lifetime: Singleton implCtor: DbQueueProcessorsService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
DbQueueProcessorsService -> DeleteDriveFilesProcessorService -> DriveService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: EndedPollNotificationProcessorService lifetime: Singleton implCtor: EndedPollNotificationProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
EndedPollNotificationProcessorService -> CreateNotificationService -> NotificationEntityService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: InboxProcessorService lifetime: Singleton implCtor: InboxProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
InboxProcessorService -> ApInboxService -> UserEntityService -> NoteEntityService -> UserEntityService
Error while validating the service descriptor 'serviceType: QueueProcessorService lifetime: Singleton implCtor: QueueProcessorService': Error: A circular dependency was detected for the service of type 'UserEntityService'.
QueueProcessorService -> SystemQueueProcessorsService -> TickChartsProcessorService -> UsersChart -> UserEntityService -> NoteEntityService -> UserEntityService
```

## Limitations
1. You need to import the `@Inject` decorator from `@/di-decorators.js` instead of `@nestjs/common`.
1. Since string values are used when resolving services, all service types must be declared by using `Symbol.for()`, and not by `Symbol()`.
1. The number of `@inject` decorators must match the number of arguments for constructor, which means you have to specify the `@Inject` decorator for every single constructor parameter, just like [vscode](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts#L233). However, this behavior may change in the future, because currently the stage 3 decorators don't support parameter decorators.

## Considerations
1. Yohira hasn't been battle-tested.
1. There might be some performance penalty on first resolve, but it should be faster once cached. We would have to run some benchmarks.
1. A wrong type could be injected accidentally by specifying a wrong type for the `@Inject` decorator, but the same goes for NestJS DI container.

By the way, is there any particular reason to use *singleton* scopes only, and not *request-scoped* and *transient* scopes?

## TODOs
- [ ] Find a better way (architecture) to resolve circular dependencies
- [ ] Fix unit and e2e tests
